"Error loading BeanIntrospection" during Jackson serialization

### Expected Behavior

In implementing the new graphql-ws protocol support in micronaut-graphql, we need to exchange various JSON message types with the client that are specified by the protocol. I created a hierarchy of `Message` classes to match the JSON message types. These types should all be serializable to JSON, using standard Jackson databind annotations for the mapping.

The message types as I would prefer to structure them are all defined here in this micronaut-graphql branch: 
https://github.com/micronaut-projects/micronaut-graphql/blob/SerializationError/graphql/src/main/java/io/micronaut/configuration/graphql/ws/Message.java

Note that sealed classes are being used to define the type structure.
 

### Actual Behaviour

When executing any of the features of the [GraphQLWsHandlerSpec](https://github.com/micronaut-projects/micronaut-graphql/blob/SerializationError/graphql/src/test/groovy/io/micronaut/configuration/graphql/ws/GraphQLWsHandlerSpec.groovy) that test handling of the [SubscribeMessage](https://github.com/micronaut-projects/micronaut-graphql/blob/e91ef4ef99fc48117189037e98115d757073c18b/graphql/src/main/java/io/micronaut/configuration/graphql/ws/Message.java#L290) type, serialization from the test client fails with the following error:

```
Error loading BeanIntrospection for type [class io.micronaut.configuration.graphql.ws.SubscribeMessage]: Bad type on operand stack
Exception Details:
  Location:
    io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection.dispatchOne(ILjava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; @106: invokespecial
  Reason:
    Type 'java/lang/Object' (current frame, stack[3]) is not assignable to 'io/micronaut/configuration/graphql/ws/SubscribeMessage$SubscribePayload'
  Current Frame:
    bci: @106
    flags: { }
    locals: { 'io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection', integer, 'java/lang/Object', 'java/lang/Object', top, 'io/micronaut/configuration/graphql/ws/SubscribeMessage' }
    stack: { uninitialized 93, uninitialized 93, 'java/lang/String', 'java/lang/Object' }
  Bytecode:
    0000000: 1baa 0000 0000 0093 0000 0000 0000 0007
    0000010: 0000 002f 0000 0037 0000 004e 0000 0056
    0000020: 0000 006d 0000 0075 0000 0080 0000 0088
    0000030: 2cc0 0061 b600 6fb0 2cc0 0061 3a04 bb00
    0000040: 6159 1904 b600 732d c000 44b7 0076 b02c
    0000050: c000 61b6 0073 b02c c000 613a 05bb 0061
    0000060: 592d c000 3219 05b6 006f b700 76b0 2cc0
    0000070: 0061 b600 79b0 bb00 7b59 127d b700 80bf
    0000080: bf2c c000 61b6 0084 b0bb 007b 5912 86b7
    0000090: 0080 bfbf 2a1b b600 8abf               
  Stackmap Table:
    same_frame(@48)
    same_frame(@56)
    same_frame(@79)
    same_frame(@87)
    same_frame(@110)
    same_frame(@118)
    full_frame(@128,{},{Object[#140]})
    full_frame(@129,{Object[#2],Integer,Object[#142],Object[#142]},{})
    same_frame(@137)
    full_frame(@147,{},{Object[#140]})
    full_frame(@148,{Object[#2],Integer,Object[#142],Object[#142]},{})

io.micronaut.core.beans.exceptions.IntrospectionException: Error loading BeanIntrospection for type [class io.micronaut.configuration.graphql.ws.SubscribeMessage]: Bad type on operand stack
Exception Details:
  Location:
    io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection.dispatchOne(ILjava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; @106: invokespecial
  Reason:
    Type 'java/lang/Object' (current frame, stack[3]) is not assignable to 'io/micronaut/configuration/graphql/ws/SubscribeMessage$SubscribePayload'
  Current Frame:
    bci: @106
    flags: { }
    locals: { 'io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection', integer, 'java/lang/Object', 'java/lang/Object', top, 'io/micronaut/configuration/graphql/ws/SubscribeMessage' }
    stack: { uninitialized 93, uninitialized 93, 'java/lang/String', 'java/lang/Object' }
  Bytecode:
    0000000: 1baa 0000 0000 0093 0000 0000 0000 0007
    0000010: 0000 002f 0000 0037 0000 004e 0000 0056
    0000020: 0000 006d 0000 0075 0000 0080 0000 0088
    0000030: 2cc0 0061 b600 6fb0 2cc0 0061 3a04 bb00
    0000040: 6159 1904 b600 732d c000 44b7 0076 b02c
    0000050: c000 61b6 0073 b02c c000 613a 05bb 0061
    0000060: 592d c000 3219 05b6 006f b700 76b0 2cc0
    0000070: 0061 b600 79b0 bb00 7b59 127d b700 80bf
    0000080: bf2c c000 61b6 0084 b0bb 007b 5912 86b7
    0000090: 0080 bfbf 2a1b b600 8abf               
  Stackmap Table:
    same_frame(@48)
    same_frame(@56)
    same_frame(@79)
    same_frame(@87)
    same_frame(@110)
    same_frame(@118)
    full_frame(@128,{},{Object[#140]})
    full_frame(@129,{Object[#2],Integer,Object[#142],Object[#142]},{})
    same_frame(@137)
    full_frame(@147,{},{Object[#140]})
    full_frame(@148,{Object[#2],Integer,Object[#142],Object[#142]},{})

	at app//io.micronaut.core.beans.DefaultBeanIntrospector.findIntrospection(DefaultBeanIntrospector.java:104)
	at app//io.micronaut.jackson.modules.BeanIntrospectionModule.findIntrospection(BeanIntrospectionModule.java:170)
	at app//io.micronaut.jackson.modules.BeanIntrospectionModule$BeanIntrospectionSerializerModifier.updateBuilder(BeanIntrospectionModule.java:302)
	at app//com.fasterxml.jackson.databind.ser.BeanSerializerFactory.constructBeanOrAddOnSerializer(BeanSerializerFactory.java:468)
	at app//com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanOrAddOnSerializer(BeanSerializerFactory.java:295)
	at app//com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:240)
	at app//com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:174)
	at app//com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1503)
	at app//com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1451)
	at app//com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:556)
	at app//com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:834)
	at app//com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:307)
	at app//com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4719)
	at app//com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3923)
	at app//io.micronaut.jackson.databind.JacksonDatabindMapper.writeValue(JacksonDatabindMapper.java:194)
	at app//io.micronaut.json.codec.MapperMediaTypeCodec.encode(MapperMediaTypeCodec.java:231)
	at app//io.micronaut.json.codec.MapperMediaTypeCodec.encode(MapperMediaTypeCodec.java:280)
	at app//io.micronaut.http.netty.websocket.WebSocketMessageEncoder.encodeMessage(WebSocketMessageEncoder.java:74)
	at app//io.micronaut.http.netty.websocket.NettyWebSocketSession.sendSync(NettyWebSocketSession.java:179)
	at app//io.micronaut.websocket.interceptor.ClientWebSocketInterceptor.send(ClientWebSocketInterceptor.java:112)
	at app//io.micronaut.websocket.interceptor.ClientWebSocketInterceptor.intercept(ClientWebSocketInterceptor.java:89)
	at app//io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at io.micronaut.configuration.graphql.ws.GraphQLWsHandlerSpec.client subscribes to a single-result query operation and receives result(GraphQLWsHandlerSpec.groovy:101)
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection.dispatchOne(ILjava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; @106: invokespecial
  Reason:
    Type 'java/lang/Object' (current frame, stack[3]) is not assignable to 'io/micronaut/configuration/graphql/ws/SubscribeMessage$SubscribePayload'
  Current Frame:
    bci: @106
    flags: { }
    locals: { 'io/micronaut/configuration/graphql/ws/$SubscribeMessage$Introspection', integer, 'java/lang/Object', 'java/lang/Object', top, 'io/micronaut/configuration/graphql/ws/SubscribeMessage' }
    stack: { uninitialized 93, uninitialized 93, 'java/lang/String', 'java/lang/Object' }
  Bytecode:
    0000000: 1baa 0000 0000 0093 0000 0000 0000 0007
    0000010: 0000 002f 0000 0037 0000 004e 0000 0056
    0000020: 0000 006d 0000 0075 0000 0080 0000 0088
    0000030: 2cc0 0061 b600 6fb0 2cc0 0061 3a04 bb00
    0000040: 6159 1904 b600 732d c000 44b7 0076 b02c
    0000050: c000 61b6 0073 b02c c000 613a 05bb 0061
    0000060: 592d c000 3219 05b6 006f b700 76b0 2cc0
    0000070: 0061 b600 79b0 bb00 7b59 127d b700 80bf
    0000080: bf2c c000 61b6 0084 b0bb 007b 5912 86b7
    0000090: 0080 bfbf 2a1b b600 8abf               
  Stackmap Table:
    same_frame(@48)
    same_frame(@56)
    same_frame(@79)
    same_frame(@87)
    same_frame(@110)
    same_frame(@118)
    full_frame(@128,{},{Object[#140]})
    full_frame(@129,{Object[#2],Integer,Object[#142],Object[#142]},{})
    same_frame(@137)
    full_frame(@147,{},{Object[#140]})
    full_frame(@148,{Object[#2],Integer,Object[#142],Object[#142]},{})

	at app//io.micronaut.core.beans.DefaultBeanIntrospector.lambda$findIntrospection$0(DefaultBeanIntrospector.java:95)
	at java.base@17.0.6/java.util.Optional.map(Optional.java:260)
	at app//io.micronaut.core.beans.DefaultBeanIntrospector.findIntrospection(DefaultBeanIntrospector.java:91)
	... 22 more
```

I have tried various solutions such as adding `@Introspected` annotations, and relocating `SubscribePayload` to a top-level class, but they do not make a difference. I ultimately had to work around the issue by using a `Map<String, Object>` for the payload of the `SubscribeMessage` instead of the `SubscribePayload` inner class. 

### Steps To Reproduce

1. Pull the `SerializationError` branch from [micronaut-graphql](https://github.com/micronaut-projects/micronaut-graphql/tree/SerializationError)
2. Execute the `GraphQLWsHandlerSpec`

### Environment Information

- macOS Monterey 12.5
- Tested with both GraaalVM CE 17.0.7 and Temurin 17.0.7

### Example Application

https://github.com/micronaut-projects/micronaut-graphql/tree/SerializationError

### Version

4.0.0.RC3

--- a/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
+++ b/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
@@ -1007,12 +1007,12 @@ final class BeanIntrospectionWriter extends AbstractAnnotationMetadataWriter {
                     Object member;
                     ClassElement propertyType;
                     DispatchWriter.DispatchTarget dispatchTarget = dispatchWriter.getDispatchTargets().get(readDispatchIndex);
-                    if (dispatchTarget instanceof DispatchWriter.MethodDispatchTarget) {
-                        MethodElement methodElement = ((DispatchWriter.MethodDispatchTarget) dispatchTarget).getMethodElement();
+                    if (dispatchTarget instanceof DispatchWriter.MethodDispatchTarget methodDispatchTarget) {
+                        MethodElement methodElement = methodDispatchTarget.getMethodElement();
                         propertyType = methodElement.getGenericReturnType();
                         member = methodElement;
-                    } else if (dispatchTarget instanceof DispatchWriter.FieldGetDispatchTarget) {
-                        FieldElement field = ((DispatchWriter.FieldGetDispatchTarget) dispatchTarget).getField();
+                    } else if (dispatchTarget instanceof DispatchWriter.FieldGetDispatchTarget fieldGetDispatchTarget) {
+                        FieldElement field = fieldGetDispatchTarget.getField();
                         propertyType = field.getGenericType();
                         member = field;
                     } else {
@@ -1043,20 +1043,27 @@ final class BeanIntrospectionWriter extends AbstractAnnotationMetadataWriter {
                     for (int i = 0; i < parameters.length; i++) {
                         ParameterElement parameter = parameters[i];
                         Object constructorArgument = constructorArguments[i];
+                        boolean isPrimitive;
                         if (constructorArgument == this) {
                             constructorWriter.loadArg(2);
+                            isPrimitive = false;
                         } else if (constructorArgument instanceof MethodElement readMethod) {
                             constructorWriter.loadLocal(prevBeanTypeLocal, beanType);
                             invokeMethod(constructorWriter, readMethod);
+                            isPrimitive = readMethod.getReturnType().isPrimitive();
                         } else if (constructorArgument instanceof FieldElement fieldElement) {
                             constructorWriter.loadLocal(prevBeanTypeLocal, beanType);
                             invokeGetField(constructorWriter, fieldElement);
+                            isPrimitive = fieldElement.isPrimitive();
                         } else {
                             throw new IllegalStateException();
                         }
-                        pushCastToType(constructorWriter, parameter);
-                        if (!parameter.isPrimitive()) {
-                            pushBoxPrimitiveIfNecessary(parameter, constructorWriter);
+                        if (isPrimitive) {
+                            if (!parameter.isPrimitive()) {
+                                pushBoxPrimitiveIfNecessary(parameter, constructorWriter);
+                            }
+                        } else {
+                            pushCastToType(constructorWriter, parameter);
                         }
                     }
                 });

--- a/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
+++ b/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
@@ -1045,20 +1045,18 @@ final class BeanIntrospectionWriter extends AbstractAnnotationMetadataWriter {
                         Object constructorArgument = constructorArguments[i];
                         if (constructorArgument == this) {
                             constructorWriter.loadArg(2);
-                            pushCastToType(constructorWriter, parameter);
-                            if (!parameter.isPrimitive()) {
-                                pushBoxPrimitiveIfNecessary(parameter, constructorWriter);
-                            }
-                        } else if (constructorArgument instanceof MethodElement) {
-                            MethodElement readMethod = (MethodElement) constructorArgument;
+                        } else if (constructorArgument instanceof MethodElement readMethod) {
                             constructorWriter.loadLocal(prevBeanTypeLocal, beanType);
                             invokeMethod(constructorWriter, readMethod);
-                        } else if (constructorArgument instanceof FieldElement) {
-                            FieldElement fieldElement = (FieldElement) constructorArgument;
+                        } else if (constructorArgument instanceof FieldElement fieldElement) {
                             constructorWriter.loadLocal(prevBeanTypeLocal, beanType);
                             invokeGetField(constructorWriter, fieldElement);
                         } else {
                             throw new IllegalStateException();
+                        }
+                        pushCastToType(constructorWriter, parameter);
+                        if (!parameter.isPrimitive()) {
+                            pushBoxPrimitiveIfNecessary(parameter, constructorWriter);
                         }
                     }
                 });

--- a/inject-java/src/test/groovy/io/micronaut/inject/beans/concopy/ConstructorCopySpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/beans/concopy/ConstructorCopySpec.groovy
@@ -1,0 +1,15 @@
+package io.micronaut.inject.beans.concopy
+
+import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
+import io.micronaut.core.beans.BeanIntrospection
+
+class ConstructorCopySpec extends AbstractTypeElementSpec {
+
+    void "test constructor dispatch"() {
+        when:
+            BeanIntrospection.getIntrospection(SubscribeMessage.class)
+        then:
+            noExceptionThrown()
+    }
+
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/beans/concopy/Message.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/beans/concopy/Message.java
@@ -1,0 +1,493 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.inject.beans.concopy;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.micronaut.core.annotation.Introspected;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.util.StringUtils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A class for mapping graphql-ws messages.
+ *
+ * @author Jeremy Grelle
+ * @since 4.0
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @Type(value = ConnectionInitMessage.class, name = Message.Types.CONNECTION_INIT),
+    @Type(value = ConnectionAckMessage.class, name = Message.Types.CONNECTION_ACK),
+    @Type(value = PingMessage.class, name = Message.Types.PING),
+    @Type(value = PongMessage.class, name = Message.Types.PONG),
+    @Type(value = SubscribeMessage.class, name = Message.Types.SUBSCRIBE),
+    @Type(value = ErrorMessage.class, name = Message.Types.ERROR),
+    @Type(value = CompleteMessage.class, name = Message.Types.COMPLETE)
+})
+@Introspected
+public abstract sealed class Message {
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @JsonIgnore
+    @NonNull
+    abstract String getMessageType();
+
+    /**
+     * Validate a required message id.
+     *
+     * @param id The required message id
+     */
+    protected void checkRequiredId(String id) {
+        if (StringUtils.isEmpty(id)) {
+            throw new IllegalArgumentException("'id' is required for messages with type '" + getMessageType() + "'.");
+        }
+    }
+
+    /**
+     * The allowable graphql-ws message types.
+     */
+    static final class Types {
+
+        static final String CONNECTION_INIT = "connection_init";
+        static final String CONNECTION_ACK = "connection_ack";
+        static final String PING = "ping";
+        static final String PONG = "pong";
+        static final String SUBSCRIBE = "subscribe";
+        static final String NEXT = "next";
+        static final String ERROR = "error";
+        static final String COMPLETE = "complete";
+
+        private Types() { }
+    }
+}
+
+/**
+ * A graphql-ws message that contains an optional payload.
+ *
+ * @param <T> The payload type
+ */
+abstract sealed class PayloadMessage<T> extends Message {
+
+    @Nullable
+    private final T payload;
+
+    /**
+     * Default constructor for a graphql-ws message with an optional payload.
+     */
+    protected PayloadMessage() {
+        this(null);
+    }
+
+    /**
+     * Constructor for a graphql-ws message with a payload.
+     *
+     * @param payload The message payload.
+     */
+    protected PayloadMessage(@Nullable T payload) {
+        this.payload = payload;
+    }
+
+    /**
+     * Get the message payload.
+     *
+     * @return The message payload.
+     */
+    @Nullable
+    public T getPayload() {
+        return payload;
+    }
+}
+
+/**
+ * A graphql-ws message that has a required non-null payload.
+ *
+ * @param <T> The payload type.
+ */
+abstract sealed class RequiredPayloadMessage<T> extends Message {
+
+    @NonNull
+    private final T payload;
+
+    /**
+     * Constructor for a graphql-ws message with a required payload.
+     *
+     * @param payload The message payload.
+     */
+    protected RequiredPayloadMessage(@NonNull T payload) {
+        Objects.requireNonNull(payload, "A payload is required for message type '" + getMessageType() + ".");
+        this.payload = payload;
+    }
+
+    /**
+     * Get the message payload - will never be <code>null</code>.
+     *
+     * @return The message payload.
+     */
+    @NonNull
+    public T getPayload() {
+        return payload;
+    }
+}
+
+/**
+ * A graphql-ws message for connection initialisation.
+ */
+final class ConnectionInitMessage extends PayloadMessage<Map<String, Object>> {
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.CONNECTION_INIT;
+    }
+}
+
+/**
+ * A graphql-ws message for connection acknowledgement.
+ */
+final class ConnectionAckMessage extends PayloadMessage<Map<String, Object>> {
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.CONNECTION_ACK;
+    }
+}
+
+/**
+ * A graphql-ws message for a ping.
+ */
+final class PingMessage extends PayloadMessage<Map<String, Object>> {
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.PING;
+    }
+}
+
+/**
+ * A graphql-ws message for a pong.
+ */
+final class PongMessage extends PayloadMessage<Map<String, Object>> {
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.PONG;
+    }
+}
+
+/**
+ * A graphql-ws message for subscribing to the execution of a query.
+ */
+@Introspected
+final class SubscribeMessage extends RequiredPayloadMessage<SubscribeMessage.SubscribePayload> {
+
+    @NonNull
+    private final String id;
+
+    /**
+     * Constructor for a graphql-ws 'subscribe' message.
+     *
+     * @param id
+     * @param payload
+     */
+    @JsonCreator
+    public SubscribeMessage(@NonNull @JsonProperty("id") String id, @NonNull @JsonProperty("payload") SubscribePayload payload) {
+        super(payload);
+        checkRequiredId(id);
+        this.id = id;
+    }
+
+    /**
+     * Get the required non-empty message id.
+     *
+     * @return The message id.
+     */
+    @NonNull
+    public String getId() {
+        return id;
+    }
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.SUBSCRIBE;
+    }
+
+    /**
+     * Get the message payload as a {@link SubscribePayload}.
+     *
+     * @return The message payload.
+     */
+    @JsonIgnore
+    @NonNull
+    SubscribePayload getSubscribePayload() {
+        //return SubscribePayload.fromMap(getPayload());
+        return getPayload();
+    }
+
+    @Introspected
+    public static class SubscribePayload {
+
+        @NonNull
+        private final String query;
+
+        @Nullable
+        private String operationName;
+
+        @Nullable
+        private Map<String, Object> variables;
+
+        @Nullable
+        private Map<String, Object> extensions;
+
+        /**
+         * Constructor for a graphql-ws 'subscribe' message's payload.
+         *
+         * @param query The required non-empty query being executed and subscribed.
+         */
+        @JsonCreator
+        public SubscribePayload(@NonNull @JsonProperty("query") String query) {
+            if (StringUtils.isEmpty(query)) {
+                throw new IllegalArgumentException("The 'query' field is required in the payload of message type '" + Types.SUBSCRIBE + "'");
+            }
+            this.query = query;
+        }
+
+        private SubscribePayload(@NonNull String query, @Nullable String operationName, @Nullable Map<String, Object> variables, @Nullable Map<String, Object> extensions) {
+            this.query = query;
+            this.operationName = operationName;
+            this.variables = variables;
+            this.extensions = extensions;
+        }
+
+        /**
+         * Gets the required non-empty query field of the message payload.
+         *
+         * @return The query.
+         */
+        @NonNull
+        public String getQuery() {
+            return query;
+        }
+
+        /**
+         * Gets the operation name of the payload.
+         *
+         * @return The operation name.
+         */
+        @Nullable
+        public String getOperationName() {
+            return operationName;
+        }
+
+        /**
+         * Sets the operation name of the payload.
+         *
+         * @param operationName The operation name.
+         */
+        public void setOperationName(@Nullable String operationName) {
+            this.operationName = operationName;
+        }
+
+        /**
+         * Gets the variables of the payload.
+         *
+         * @return The payload variables.
+         */
+        @Nullable
+        public Map<String, Object> getVariables() {
+            return variables;
+        }
+
+        /**
+         * Sets the variables of the payload.
+         *
+         * @param variables The operation name.
+         */
+        public void setVariables(@Nullable Map<String, Object> variables) {
+            this.variables = variables;
+        }
+
+        /**
+         * Gets the extensions of the payload.
+         *
+         * @return The extensions.
+         */
+        @Nullable
+        public Map<String, Object> getExtensions() {
+            return extensions;
+        }
+
+        /**
+         * Sets the extensions of the payload.
+         *
+         * @param extensions The operation name.
+         */
+        public void setExtensions(@Nullable Map<String, Object> extensions) {
+            this.extensions = extensions;
+        }
+
+        public Map<String, Object> toMap() {
+            return Map.of(
+                "query", this.query,
+                "operationName", Optional.ofNullable(this.operationName).orElse(""),
+                "variables", Optional.ofNullable(this.variables).orElse(new HashMap<>()),
+                "extensions", Optional.ofNullable(this.extensions).orElse(new HashMap<>())
+            );
+        }
+
+        @SuppressWarnings("unchecked")
+        public static SubscribePayload fromMap(Map<String, Object> payload) {
+            return new SubscribePayload(
+                (String) payload.get("query"),
+                (String) payload.get("operationName"),
+                (Map<String, Object>) payload.get("variables"),
+                (Map<String, Object>) payload.get("extensions"));
+        }
+    }
+}
+
+/**
+ * A graphql-ws message for reporting errors pertaining to a specific subscription.
+ */
+final class ErrorMessage extends RequiredPayloadMessage<List<Map<String, Object>>> {
+
+    @NonNull
+    private final String id;
+
+    /**
+     * Constructor for a graphql-ws 'error' message.
+     *
+     * @param id     The required non-empty id of the message.
+     * @param errors The errors resulting from the specific subscription with the corresponding id.
+     */
+    @JsonCreator
+    public ErrorMessage(@NonNull @JsonProperty("id") String id, @NonNull @JsonProperty("payload") List<Map<String, Object>> errors) {
+        super(errors);
+        checkRequiredId(id);
+        this.id = id;
+    }
+
+    /**
+     * Get the required non-empty message id.
+     *
+     * @return The message id.
+     */
+    @NonNull
+    public String getId() {
+        return id;
+    }
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.ERROR;
+    }
+}
+
+/**
+ * A graphql-ws message for the completion of a subscription.
+ */
+final class CompleteMessage extends Message {
+
+    @NonNull
+    private final String id;
+
+    /**
+     * Constructor for a graphql-ws 'complete' message.
+     *
+     * @param id The required non-empty id of the message.
+     */
+    public CompleteMessage(@NonNull String id) {
+        checkRequiredId(id);
+        this.id = id;
+    }
+
+    /**
+     * Get the required non-empty message id.
+     *
+     * @return The message id.
+     */
+    @NonNull
+    public String getId() {
+        return id;
+    }
+
+    /**
+     * Get the required value of the message's <code>type</code> field.
+     *
+     * @return The message's type
+     */
+    @Override
+    @JsonIgnore
+    @NonNull
+    String getMessageType() {
+        return Types.COMPLETE;
+    }
+}

